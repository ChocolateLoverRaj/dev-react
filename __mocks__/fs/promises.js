/* global jest */

// Mocker for fs/promises

const fs = jest.createMockFromModule('fs/promises')

const EventEmitter = require('eventemitter3')

const frozen = new Set()
fs._frozen = frozen
const errorFiles = new Set()
fs._errorFiles = errorFiles
const files = new Map()
fs._files = files
fs._reset = () => {
  frozen.clear()
  errorFiles.clear()
  files.clear()
}

const mock = new EventEmitter()
  .on('freeze', filename => {
    frozen.add(filename)
  })
  .on('unfreeze', filename => {
    frozen.delete(filename)
  })
  .on('throw', filename => {
    errorFiles.add(filename)
  })
  .on('reset', () => {
    fs._reset()
  })
fs._mock = mock

const onceUnfrozen = filename => new Promise((resolve, reject) => {
  const handler = f => {
    if (f === filename) {
      mock.off(handler)
      resolve()
    }
  }
  mock.on('unfreeze', handler)
})

fs.writeFile = async (filename, content) => {
  if (errorFiles.has(filename)) {
    throw new Error('Error writing file.')
  }
  const write = () => {
    files.set(filename, content)
    mock.emit('wrote', filename, content)
  }
  if (frozen.has(filename)) {
    await onceUnfrozen(filename)
    write()
  } else {
    write()
  }
}

fs.readFile = async filename => {
  if (errorFiles.has(filename)) {
    throw new Error('Error reading file.')
  }
  const read = () => {
    mock.emit('read', filename)
    return files.get(filename)
  }
  if (frozen.has(filename)) {
    await onceUnfrozen(filename)
    return read()
  } else {
    return read()
  }
}

fs.unlink = async filename => {
  files.delete(filename)
  mock.emit('unlinked', filename)
}

module.exports = fs
