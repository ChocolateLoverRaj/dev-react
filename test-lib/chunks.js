const chunks = stream => new Promise((resolve, reject) => {
  const chunks = []
  stream.on('data', data => {
    chunks.push(data)
  })
  stream.on('end', () => {
    resolve(chunks)
  })
  stream.on('error', reject)
})

module.exports = chunks
