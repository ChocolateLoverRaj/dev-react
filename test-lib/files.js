export class File {
  constructor () {
    this.canRead = true
    this.canWrite = true
  }
}

export class NormalFile extends File {
  constructor (content) {
    super()
    this.content = content
  }
}

export class Dir extends File {
  constructor () {
    super()
    this.files = new Map()
  }
}

class EnoentError extends Error {
  constructor () {
    super(...arguments)
    this.code = 'ENOENT'
  }
}

export const topDir = new Dir()

export const reset = () => {
  topDir.files.clear()
}

export const getFile = path => {
  let file = topDir
  const filenames = path.split('/')
  let fileTree = []
  for (const filename of filenames) {
    fileTree.push(filename)
    if (file.files.has(filename)) {
      file = file.files.get(filename)
    } else {
      throw new EnoentError(`File: ${fileTree.join('/')} does not exist`)
    }
  }
  return file
}

export const setFile = (path, fileToSet) => {
  let dir = topDir
  const filenames = path.split('/')
  let fileTree = []
  for (const filename of filenames) {
    fileTree.push(filename)
    if (fileTree.length === filenames.length) {
      dir.files.set(filename, fileToSet)
      return
    }
    if (dir.files.has(filename)) {
      dir = dir.files.get(filename)
    } else {
      throw new EnoentError(`File: ${fileTree.join('/')} does not exist`)
    }
  }
}

export const unlinkFile = path => {
  let file = topDir
  const filenames = path.split('/')
  let fileTree = []
  for (const filename of filenames) {
    fileTree.push(filename)
    if (file.files.has(filename)) {
      if (fileTree.length === filenames.length) {
        file.files.delete(filename)
        return
      }
      file = file.files.get(filename)
    } else {
      throw new EnoentError(`File: ${fileTree.join('/')} does not exist`)
    }
  }
}
