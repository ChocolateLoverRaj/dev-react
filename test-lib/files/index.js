// TODO: structure with folders and files
const topDir = new Map()

export default topDir