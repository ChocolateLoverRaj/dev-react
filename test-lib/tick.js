// Returns a promise that gets resolved after process.nextTick()

const tick = () => new Promise(resolve => {
  process.nextTick(resolve)
})

module.exports = tick
