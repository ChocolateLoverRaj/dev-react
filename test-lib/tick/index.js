// Returns a promise that gets resolved after process.nextTick()

const tick = () => new Promise(resolve => {
  process.nextTick(resolve)
})

export default tick
