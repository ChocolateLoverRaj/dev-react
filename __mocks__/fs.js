// Mocker for fs
import { getFile } from '../test-lib/files.js'
import tick from '../test-lib/tick.js'
import { Readable, PassThrough } from 'stream'

export const _readStreamErrorFiles = new Set()

export const _reset = () => {
  _readStreamErrorFiles.clear()
}

export const createReadStream = path => {
  if (_readStreamErrorFiles.has(path)) {
    const stream = new PassThrough()
    tick().then(() => {
      stream.destroy(new Error('Error creating a read stream.'))
    })
    return stream
  }
  return Readable.from(getFile(path).content.split(''))
}
