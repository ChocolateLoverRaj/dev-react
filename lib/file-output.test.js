/* global afterEach, describe, test, expect */
import FileOutput from './file-output.js'

import * as fsPromises from 'fs/promises'
import * as fs from 'fs'
import { getFile } from '../test-lib/files.js'
import noResolve from '../test-lib/no-resolve.js'
import tick from '../test-lib/tick.js'
import chunks from '../test-lib/chunks.js'
import unsafeOnce from '../test-lib/unsafe-once.js'

import { once } from 'events'

const codeBuilder = async () => 'code'
const newCodeBuilder = async () => 'new code'
const errorBuilder = async () => {
  throw new Error('A build error.')
}
const outputPath = 'output'

afterEach(fsPromises._reset)
afterEach(fs._reset)

describe('static', () => {
  test('OutdatedError', () => {
    expect(new FileOutput.OutdatedError()).toMatchSnapshot()
  })

  test('DestroyedError', () => {
    expect(new FileOutput.DestroyedError()).toMatchSnapshot()
  })
})

describe('update', () => {
  describe('canceling', () => {
    test('overwrite', async () => {
      const fileOutput = new FileOutput(outputPath)
      let canceled = false
      fileOutput.update(async emitter => {
        emitter.once('cancel', () => {
          canceled = true
        })
        await noResolve(tick(), once(emitter, 'cancel'))
      })
      const update2 = fileOutput.update(codeBuilder)
      expect(canceled).toBe(true)
      await update2
    })

    test('destroy', async () => {
      const fileOutput = new FileOutput(outputPath)
      let canceled = false
      fileOutput.update(async emitter => {
        emitter.once('cancel', () => {
          canceled = true
        })
        await noResolve(tick(), once(emitter, 'cancel'))
      })
      const destroyPromise = fileOutput.destroy()
      expect(canceled).toBe(true)
      await destroyPromise
    })
  })

  test('write', async () => {
    const fileOutput = new FileOutput(outputPath)
    await fileOutput.update(codeBuilder)
    expect(getFile(outputPath).content).toBe('code')
    expect(fileOutput.buildPromise).toBe(null)
  })

  describe('overwrite', () => {
    test('mid-build', async () => {
      const builder = async () => {
        await tick()
        return 'code'
      }
      const fileOutput = new FileOutput(outputPath)
      fileOutput.update(builder)
      await tick()
      const update2 = fileOutput.update(newCodeBuilder)
      expect((await once(fsPromises._mock, 'wrote'))[1]).toBe('new code')
      await noResolve(once(fsPromises._mock, 'wrote'), update2)
    })

    test('mid-write', async () => {
      const fileOutput = new FileOutput(outputPath)
      const update1 = fileOutput.update(codeBuilder)
      fsPromises._frozen.add(outputPath)
      await fileOutput.buildPromise
      const update2 = fileOutput.update(async () => {
        await tick()
        return 'new code'
      })
      fsPromises._mock.emit('unfreeze', outputPath)
      await update1
      expect(getFile(outputPath).content).toBe('code')
      expect(fileOutput.buildPromise).not.toBe(null)
      await update2
      expect(getFile(outputPath).content).toBe('new code')
      expect(fileOutput.buildPromise).toBe(null)
    })
  })
})

describe('read', () => {
  const newCodeArray = ['new code']
  test('file stream no destroy', async () => {
    const fileOutput = new FileOutput(outputPath)
    await fileOutput.update(codeBuilder)
    expect(await chunks(fileOutput.read())).toStrictEqual(['c', 'o', 'd', 'e'])
  })

  describe('build', () => {
    test('promise', async () => {
      const fileOutput = new FileOutput(outputPath)
      fileOutput.update(codeBuilder)
      expect(await chunks(fileOutput.read())).toStrictEqual(['code'])
    })

    test('overwrite promise', async () => {
      const fileOutput = new FileOutput('2')
      fileOutput.update(codeBuilder)
      const readStream = fileOutput.read()
      fileOutput.update(newCodeBuilder)
      expect(await chunks(readStream)).toStrictEqual(newCodeArray)
    })
  })

  describe('file stream destroy', () => {
    test('pre-stream', async () => {
      const fileOutput = new FileOutput(outputPath)
      await fileOutput.update(codeBuilder)
      const fileChunks = chunks(fileOutput.read())
      fileOutput.update(newCodeBuilder)
      expect(await fileChunks).toStrictEqual(newCodeArray)
    })

    test('mid-stream', async () => {
      const fileOutput = new FileOutput(outputPath)
      await fileOutput.update(codeBuilder)
      const stream = fileOutput.read()
      expect((await once(stream, 'data'))[0]).toBe('c')
      const update2 = fileOutput.update(newCodeBuilder)
      const errorEvent = once(stream, 'error')
      await noResolve(unsafeOnce(stream, 'data'), errorEvent)
      expect((await errorEvent)[0]).toBeInstanceOf(FileOutput.OutdatedError)
      await update2
      expect(await chunks(fileOutput.read())).toStrictEqual(['n', 'e', 'w', ' ', 'c', 'o', 'd', 'e'])
    })
  })

  describe('destroy FileOutput instance', () => {
    test('pre-stream', async () => {
      const fileOutput = new FileOutput(outputPath)
      await fileOutput.update(codeBuilder)
      const stream = fileOutput.read()
      const destroyPromise = fileOutput.destroy()
      const errorEvent = once(stream, 'error')
      await noResolve(unsafeOnce(stream, 'data'), errorEvent)
      expect((await errorEvent)[0]).toBeInstanceOf(FileOutput.DestroyedError)
      await destroyPromise
    })

    test('mid-stream', async () => {
      const fileOutput = new FileOutput(outputPath)
      await fileOutput.update(codeBuilder)
      const stream = fileOutput.read()
      expect((await once(stream, 'data'))[0]).toBe('c')
      const destroyPromise = fileOutput.destroy()
      const errorEvent = once(stream, 'error')
      await noResolve(unsafeOnce(stream, 'data'), errorEvent)
      expect((await errorEvent)[0]).toBeInstanceOf(FileOutput.DestroyedError)
      await destroyPromise
    })
  })
})

describe('destroy', () => {
  test('no errors', async () => {
    const fileOutput = new FileOutput(outputPath)
    await fileOutput.update(codeBuilder)
    getFile(outputPath)
    await fileOutput.destroy()
    expect(() => {
      getFile(outputPath)
    }).toThrowError()
  })

  test('ENOENT', async () => {
    const fileOutput = new FileOutput(outputPath)
    await fileOutput.destroy()
  })
})

describe('errors', () => {
  describe('update', () => {
    test('builder', async () => {
      const fileOutput = new FileOutput(outputPath)
      await expect(fileOutput.update(errorBuilder)).rejects.toMatchSnapshot()
    })

    test('write', async () => {
      const fileOutput = new FileOutput(outputPath)
      fsPromises._errorFiles.add(outputPath)
      await expect(fileOutput.update(codeBuilder)).rejects.toMatchSnapshot()
    })
  })

  describe('read', () => {
    describe('from build promise', () => {
      test('no overwrite', async () => {
        const fileOutput = new FileOutput(outputPath)
        fileOutput.update(errorBuilder).catch(() => { })
        await expect(chunks(fileOutput.read())).rejects.toMatchSnapshot()
      })

      test('overwrite', async () => {
        const fileOutput = new FileOutput(outputPath)
        fileOutput.update(codeBuilder)
        const readChunks = chunks(fileOutput.read())
        fileOutput.update(errorBuilder).catch(() => { })
        await expect(readChunks).rejects.toMatchSnapshot()
      })
    })

    describe('from read stream', () => {
      test('no overwrite', async () => {
        const fileOutput = new FileOutput(outputPath)
        await fileOutput.update(codeBuilder)
        fs._readStreamErrorFiles.add(outputPath)
        await expect(chunks(fileOutput.read())).rejects.toMatchSnapshot()
      })

      test('overwrite', async () => {
        const fileOutput = new FileOutput(outputPath)
        await fileOutput.update(codeBuilder)
        const streamChunks = chunks(fileOutput.read())
        fileOutput.update(errorBuilder).catch(() => { })
        await expect(streamChunks).rejects.toMatchSnapshot()
      })
    })
  })

  test('destroy', async () => {
    const fileOutput = new FileOutput(outputPath)
    fsPromises._errorFiles.add(outputPath)
    await expect(fileOutput.destroy()).rejects.toMatchSnapshot()
  })
})
