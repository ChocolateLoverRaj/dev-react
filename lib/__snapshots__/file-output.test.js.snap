// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errors destroy 1`] = `[Error: Error unlinking file: Error: Error unlinking file.]`;

exports[`errors read from build promise no overwrite 1`] = `[Error: Error with buildPromise: Error: A build error.]`;

exports[`errors read from build promise overwrite 1`] = `[Error: Error with buildPromise: Error: A build error.]`;

exports[`errors read from read stream no overwrite 1`] = `[Error: Error with read stream: Error: Error creating a read stream.]`;

exports[`errors read from read stream overwrite 1`] = `[Error: Error with buildPromise: Error: A build error.]`;

exports[`errors update builder 1`] = `[Error: Builder error: Error: A build error.]`;

exports[`errors update write 1`] = `[Error: Error writing file: Error: Error writing file.]`;

exports[`static DestroyedError 1`] = `[Error: The FileOutput was destroyed.]`;

exports[`static OutdatedError 1`] = `[Error: A new update was created while you started streaming the file. Try again to stream the latest file.]`;
